File: go/portscan.go
package main
// Run with:
// go run portscan.go www.example.com

import (
	"time"
	"strconv"
	"log"
	"net"
	"os"
)

func printUsage() {
	log.Println("Usage: ")
	log.Println("  go run portscan.go <host> ")
	log.Println("Example: ")
	log.Println("  go run portscan.go www.example.com")
	log.Println("  go run portscan.go 8.8.8.8")
}

func testTcpConnect(host string, port int, doneChannel chan bool) {
	timeoutLength := 5 * time.Second
	conn, err := net.DialTimeout("tcp", host +":" + strconv.Itoa(port), timeoutLength)
	if err != nil {
		doneChannel <- false
		return // Could not connect
	} 
	conn.Close()
	log.Printf("[+] %d connected", port)
	doneChannel <- true
}

func main() {
	if len(os.Args) == 1 {
		log.Println("No arguments provided.")
		printUsage()
		os.Exit(1)
	}
	doneChannel := make(chan bool)
	activeThreadCount := 0
	log.Println("Scanning host: " + os.Args[1])
	for portNumber := 1; portNumber <= 65535; portNumber++ {
		activeThreadCount++
		go testTcpConnect(os.Args[1], portNumber, doneChannel)
	}

	for {
		<- doneChannel
		activeThreadCount--;
		if activeThreadCount == 0 {
			break
		}
	}
}

File: java/PortScan.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;

// Run with:
// javac PortScan.java
// java PortScan www.example.com
class PortScan {

    private static void printUsage() {
        System.out.println("Usage:");
        System.out.println("  java PortScan <host>");
        System.out.println("Examples:");
        System.out.println("  java PortScan 127.0.0.1");
        System.out.println("  java PortScan www.devdungeon.com");
    }

    private static void checkPort(String host, int portNumber) {
        Thread thread = new Thread(() -> {
            Socket clientSocket = new Socket();
            try {
                clientSocket.connect(new InetSocketAddress(host, portNumber), 3000);
                System.out.printf("[+] %d connected\n", portNumber);
                clientSocket.close();
            } catch (IOException e) {
                //System.out.printf("[-] Could not connect %d", portNumber);
                //e.printStackTrace();
            }
        });
        thread.start();
    }

    public static void main(String[] args) {
        if (args.length == 0) {
            System.out.println("No arguments received.");
            printUsage();
            System.exit(1);
        }

        String host = args[0];
        System.out.println(args);
        System.out.println(host);
        System.out.printf("Scanning %s now...\n", host);

        for (Integer portNumber = 1; portNumber <= 65535; portNumber++) {
            checkPort(host, portNumber);
        }

    }
}